nomad/alloc_endpoint.go

```diff
@@ -565,43 +565,113 @@ func (a *Alloc) SignIdentities(args *structs.AllocIdentitiesRequest, reply *stru
 			continue
 		}
 
-		task := out.LookupTask(idReq.WorkloadIdentifier)
-		if task == nil {
-			// Job has likely been updated to remove this task
-			reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
-				WorkloadIdentityRequest: *idReq,
-				Reason:                  structs.WIRejectionReasonMissingTask,
-			})
-			continue
-		}
+		job := out.Job
+
+		switch idReq.WorkloadType {
+		case structs.TaskWorkload:
+			task := out.LookupTask(idReq.WorkloadIdentifier)
+			if task == nil {
+				// Job has likely been updated to remove this task
+				reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
+					WorkloadIdentityRequest: *idReq,
+					Reason:                  structs.WIRejectionReasonMissingTask,
+				})
+				continue
+			}
 
-		widFound := false
-		for _, wid := range task.Identities {
-			if wid.Name != idReq.IdentityName {
+			widFound, err := a.signTasks(task, out, idReq, reply, now)
+			if err != nil {
+				return err
+			}
+
+			if !widFound {
+				reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
+					WorkloadIdentityRequest: *idReq,
+					Reason:                  structs.WIRejectionReasonMissingIdentity,
+				})
 				continue
 			}
 
-			widFound = true
-			claims := structs.NewIdentityClaims(out.Job, out, idReq.WorkloadIdentifier, wid, now)
-			token, _, err := a.srv.encrypter.SignClaims(claims)
+		case structs.ServiceWorkload:
+			widFound, err := a.signServices(job, out, idReq, reply, now)
 			if err != nil {
 				return err
 			}
-			reply.SignedIdentities = append(reply.SignedIdentities, &structs.SignedWorkloadIdentity{
-				WorkloadIdentityRequest: *idReq,
-				JWT:                     token,
-				Expiration:              claims.Expiry.Time(),
-			})
-			break
+
+			if !widFound {
+				reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
+					WorkloadIdentityRequest: *idReq,
+					Reason:                  structs.WIRejectionReasonMissingIdentity,
+				})
+				continue
+			}
 		}
+	}
+	return nil
+}
 
-		if !widFound {
-			reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
-				WorkloadIdentityRequest: *idReq,
-				Reason:                  structs.WIRejectionReasonMissingIdentity,
-			})
+func (a *Alloc) signTasks(
+	task *structs.Task,
+	alloc *structs.Allocation,
+	idReq *structs.WorkloadIdentityRequest,
+	reply *structs.AllocIdentitiesResponse,
+	now time.Time,
+) (widFound bool, err error) {
+	for _, wid := range task.Identities {
+		if wid.Name != idReq.IdentityName {
 			continue
 		}
+
+		widFound = true
+		err = a.signIdentities(alloc, wid, idReq, reply, now)
 	}
+	return
```

Comment:
Do we expect to only have a single match her? In case there are multiple we may lose an `err`. For example, if the first match errors, but the second succeeds `err` will return `nil`.

We could make the intention of a single match more clear by `break`in the loop on match instead of `continue`ing on mismatch, or a `return` like in `signServices` below.